src/execution/execute.js

```diff
@@ -1026,15 +1122,34 @@ function defaultResolveTypeFn(
   value: mixed,
   context: mixed,
   info: GraphQLResolveInfo,
-  abstractType: GraphQLAbstractType
-): ?GraphQLObjectType {
+  abstractType: GraphQLAbstractType,
+): ?GraphQLObjectType | ?Promise<?GraphQLObjectType> {
   const possibleTypes = info.schema.getPossibleTypes(abstractType);
+  const promisedIsTypeOfResults = [];
+  const promisedIsTypeOfResultTypes = [];
+
   for (let i = 0; i < possibleTypes.length; i++) {
     const type = possibleTypes[i];
-    if (type.isTypeOf && type.isTypeOf(value, context, info)) {
-      return type;
+
+    if (type.isTypeOf) {
+      const isTypeOfResult = type.isTypeOf(value, context, info);
+
+      if (isThenable(isTypeOfResult)) {
+        promisedIsTypeOfResults.push(isTypeOfResult);
+        promisedIsTypeOfResultTypes.push(type);
+      } else if (isTypeOfResult) {
+        return type;
+      }
     }
   }
+
+  if (promisedIsTypeOfResults.length) {
+    return Promise.all(promisedIsTypeOfResults).then(isTypeOfResults => (
+      promisedIsTypeOfResultTypes[
+        isTypeOfResults.findIndex(isTypeOf => isTypeOf)
+      ]
+    ));
+  }
```

Comment:
Re: the order of possible types: At Facebook we actually have a version of this resolver that runs in development environments (due to performance concerns) that runs through all possible types and makes an assertion that exactly one of them is the correct type.